### ISSUE TITLE
Refactoring the terminal.js module

### ISSUE THREAD

**Issue by xovertheyearsx** [foreshadow/atom-python-run#46]
**TL;DR**
i just want to make sure the code for `terminal` is readable and maintainable. not sure how to go about it. especially since i have more experience with Python than JS. i'll also be adding some documentation for `cp` and `terminal` to help others understand how the code works.

**explanation**
the `terminal.js` module is something i quickly hacked together to create cross platform capability for the `atom-python-run` module.

i'm in the midst of refactoring the `terminal.js` module and i'm just not sure how to do it and might require a live coding session.

i also have an idea for fixing the absolute path issue; that way it can be relative or absolute depending on the situation in relation to an issue mentioned in issue #44. 

so i've decided to new this as an issue hoping to resolve it. 

over the next few days, or even weeks, depending on how much time i'm able to invest in to this, i will be proposing models to help resolve it. 

these models may be pictorial or actual code that i've written in a separate branch. most of the code displayed here may or may not be used at all.

looking over the code, it's quite disturbing to look at and difficult to decipher; some of which was unfortunately unavoidable. it was either a `switch` statement or a long series of `branching` statements. either of which, the `switch` still seems to be a better choice. i'm also accustomed to thinking in terms of classes, not functional, prototypical, objects.

i'm very inexperienced in functional programming languages, so JS seems to challenge me in that regard and i'm currently not up to the challenge. after this issue is resolved, and making some minor changes or enhancements to `cp`, i've decided i'll be moving on to other ventures that will inevitably take up much of my time and attention.

before i move on, i do hope to leave behind enough information to break down both `cp` and `terminal` so that they can effectively be maintained.

**Comment by xovertheyearsx** [comment]
## terminal.js
### how it's supposed to work in theory
**coffeescript or javascript?**
atom uses coffeescript by default (imho is still a curveball) and can still use javascript because node is built-in to it by default. npm provides a coffeescript package which is used by node to translate and/or execute the code. i'm comfortably acquainted with javascript, so i wrote it in javascript.

**where i'm going with this?**
i'll start with **linux** first, then work my way to **mac os x**, and then leave **windows** for last. **linux** is the most complicated out of the 3 and requires extra information that the others do not. **mac os x** requires an extra step that windows does not. this basically means that **windows** has the most basic setup, mainly in part because of its unilateral, and universal, nature.

**what is a tty?**
the idea was to create an object named `tty`. the object `tty` is meant to handle a terminal, or console, or command line interface window. it has many names, most of which have been adapted throughout [history](https://en.wikipedia.org/wiki/Teleprinter). which is why i constantly refer to it as a tty, terminal, or cli instance; because that's exactly what it is. 

today, on a modern linux console, each active login terminal session is referred to as a `tty`. which is what i named it after. in windows, we informally refer to it as the dos-prompt, or command prompt.

**what is a tty in atom-python-run?**
the first step is to identify the operating system. after that, we determine the default shell, its execution option, and the script to be executed. in `atom-python-run`, that script happens to be `cp`. `cp` then executes a script to be passed to it. this is usually a command set in settings and looks something like this: `python {file}`. hit f5 or f6 and you create a terminal instance which calls a script within that terminal instance.

**what is a tty process?**
node has a few modules, objects, and methods for handling processes. the one used in `atom-python-run` is `child_process.spawn`. because javascript is functional (and weird with scoping) and is an event-driven language, you can only interact with a spawned process during an occurring event. it's easier to just feed it input and grab the output than to manipulate the process; which is why `cp` is used. instead of creating a view, we ask the OS to create a tty instance. `cp` executes the selected script and "holds" the instance open waiting for user input.

**Comment by xovertheyearsx** [comment]
## terminal.js
### how it works in application
**at a glance**
first we determine the OS, set the `tty` for the OS, `spawn` a `tty` that calls `cp`, and `cp` handles that process. this is all automated so we don't have to hard code anything in to the module, the plugin, or the any of its dependencies. this also allows a bit of flexibility on the programmers, and end users, part and allows us to be able to do this
```
let log = true;
let options = {
    cwd: info.dir,
    detached: true,
};
cmds = [log, options, nocp, ...cmds];
let tty = terminal.type.spawns_tty(...cmds);
tty.unref();
```
let me explain from the start. 

```terminal.type.spawns_tty(...cmds);```

`terminal` is the module name and `type` is the object. `terminal.type.spawns_tty` is a method that wraps `child_process.spawn` and executes according to the environment its in. 

all the logic that asks if this is windows, if it is windows, what shell am i using, if im using that shell, what option am i using, if im using that option, what am i executing; and so on and so forth is handled by the `terminal.type` object automatically in the background.

**linux**
linux has a variety of desktop environments and with these environments a variety of graphical terminals each with their own options as well. the list is long. i went for the most common environments and terminals. its ugly and can't really be helped.

first i ask it what desktop session it's using. this function returns the terminal name as a string. we then use the terminal to figure out what option its using. from there, we have an object that has the terminal name and option for that environment.
```
let shell = function () {
    switch (process.env.GDMSESSION) {
        // if session is using gtk
        case 'ubuntu':
        case 'ubuntu-2d':
        case 'gnome':
        case 'gnome-shell':
        case 'gnome-classic':
        case 'gnome-fallback':
        case 'cinnamon':
            return "gnome-terminal";
        // xfce session has its own terminal, xfce is gtk compatible
        case 'xfce':
            return "xfce4-terminal";
        // if session is using qt, kde and lxde are qt compatible
        case 'kde-plasma':
            return "konsole";
        case 'Lubuntu':
            return "lxterminal";
        default:
            // attempt to determine desktop session
            switch (process.env.XDG_CURRENT_DESKTOP) {
                case 'Unity':
                case 'GNOME':
                case 'X-Cinnamon':
                    return "gnome-terminal";
                case 'XFCE':
                    return "xfce4-terminal";
                case 'KDE':
                    return "konsole";
                case 'LXDE':
                    return "lxterminal";
                // if unknown session, default to xterm
                default:
                    return "xterm";
            }
    }
};


let shell_option = function (shell) {
    // determine the shells default execution option
    switch (shell) {
        case 'gnome-terminal':
        case 'xfce4-terminal':
            return "-x"
        case 'konsole':
        case 'lxterminal':
        default:
            return "-e"
    }
};


let Terminal = function (name, option) {
    this.name = name;
    this.option = option;
};


let name = shell();
let option = shell_option(name);

exports.terminal = new Terminal(name, option);
```
linux is the only environment that we do this for. darwin uses `quartz` and windows uses `aero`. darwin happens to use the `terminal.app` while windows uses `cmd` or `powershell`. the newer version of windows 10 with the fall creators update will officially include windows subsystem for linux (WSL) as well.

**darwin/mac os x**
darwin uses quartz desktop environment and a utility named `osascript` to execute scripts within a terminal. this means the command is passed to `osascript` and is directly passed to `terminal.app`. this is more clearly explained with the `terminal.type.has_a_tty` method used as an example.

**Comment by xovertheyearsx** [comment]
## terminal.js
### how to determine if we have a tty
**has_a_script**
this is a function that determines the default script to execute during the tty instantiation. if no filepath is given, it determines the default path to use, which is an absolute path. this allows us to use a relative path instead if we need to. this is as much for flexibility as it is dependability.
```
let has_a_script = function (filepath) {
    if (null == filepath) {
        if ('win32' === process.platform) {
            filepath = `${process.env.USERPROFILE}\\.atom\\packages\\atom-python-run\\cp\\main.py`;
        } else {
            filepath =  `${process.env.HOME}/.atom/packages/atom-python-run/cp/main.py`;
        }
    }
    return filepath;
};
```

**has_a_tty**
the general outline is that 1 of 3 objects will be created. only 1 object is ever created. each object has the general outline of having the the `shell`, `option`, `call`, and `script` to be executed by `child_process.spawn`. 

```
let Win32TTY = function (filepath) {
    this.shell = process.env.COMSPEC;
    this.option = ['/c', 'start'];
    this.call = 'python';
    this.script = has_a_script(filepath);
};
```

this object is a model of how to create a base tty instance within that operating systems environment. the only instance in which we add an extra "member" to the model is when we happen to be executing in darwin and need to call the anonymous function called `command`.

```
let DarwinTTY = function (filepath) {
    this.shell = 'osascript';
    this.option = '-e';
    this.call = 'python'
    this.script = has_a_script(filepath);
    this.command = function (...args) {
        let path = this.call + " " + this.script;
        for (let token of args) {
            path += ` ${token}`;
        }
        return `tell app "Terminal" to do script "${path}"`
    };
};
```
instead of calling `shell`, `option`, `call`, and `script` individually, we allow `command` to carefully construct what `osascript` expects; a string that is executed in a specific format. it has to be passed this way, otherwise it will fail. i tried many methods before coming to this conclusion.

**the bigger picture**
i explain this is more detail while explaining the spawn wrapper set. for now, this is what `terminal.type.has_a_tty` looks like as a general model.
```
const linux = require('./linux');


let has_a_script = function (filepath) {
    if (null == filepath) {
        if ('win32' === process.platform) {
            filepath = `${process.env.USERPROFILE}\\.atom\\packages\\atom-python-run\\cp\\main.py`;
        } else {
            filepath =  `${process.env.HOME}/.atom/packages/atom-python-run/cp/main.py`;
        }
    }
    return filepath;
};


let Win32TTY = function (filepath) {
    this.shell = process.env.COMSPEC;
    this.option = ['/c', 'start'];
    this.call = 'python';
    this.script = has_a_script(filepath);
};


let LinuxTTY = function (filepath) {
    this.shell = linux.terminal.name;
    this.option = [linux.terminal.option];
    this.call = 'python';
    this.script = has_a_script(filepath);
};


let DarwinTTY = function (filepath) {
    this.shell = 'osascript';
    this.option = '-e';
    this.call = 'python'
    this.script = has_a_script(filepath);
    this.command = function (...args) {
        let path = this.call + " " + this.script;
        for (let token of args) {
            path += ` ${token}`;
        }
        return `tell app "Terminal" to do script "${path}"`
    };
};


let has_a_tty = function (filepath) {
    switch (process.platform) {
        case 'win32':
            return new Win32TTY(filepath);
        case 'linux':
            return new LinuxTTY(filepath);
        case 'darwin':
            return new DarwinTTY(filepath);
    }
};


exports.has_a_tty = has_a_tty;
```

**Comment by xovertheyearsx** [comment]
The code is pretty much done. I completely re-factored `terminal.js`. I also fixed and touched up `atom-python-run.js`. I've also added some features to hopefully make it more flexible. 

I'll be working on the wiki over the next week or so. It will break `atom-python-run`, `terminal`, and `cp` in detail so anyone interested in understanding how it works can play around with it, modifiy it, or even fix or change anything that might need it.

**Comment by xovertheyearsx** [comment]
Regrettably, I'm not sure if I'll have enough time over the course of the next few months as I have to focus my attention else where. 

I've decided to push the existing branch upstream for the sake of having it available to others that might want to use it, fix it, or complete it. It works, for the most part, however, it is not ready for production use; the public repo should default to the current version instead.

It took much longer than expected. In my efforts to refactoring the core code, I realized a fundamental flaw in the design which was mistakenly added to it (switching `cp` on and off). This led me to refactoring `cp` core code as well for parsing arguments passed to it. 

Instead, it should have the ability to switch pausing on and off. This allows `cp` to log and filter anything that may be executed by it (for security reasons). Otherwise, anything can be executed by passing it to `atom` or `atom-python-run` directly which could end up being very bad.

Most of the reasons I won't be available are due to personal issues I'm going through lately. I've enjoyed collaborating on this project and hope to continue to do so in the near future. I'll also be keeping an eye on this project seeing as I do use it during my development cycles.

I'm not sure when I will push the refactored code; I am hoping to do so by the end of October at the latest to see that it maintains a bit of coherency.

**Comment by Megamiun** [comment]
Hello, @xovertheyearsx.

Is there anyway you can push your code to another branch? I was trying to use this plugin on my Mint with MATE DE and it was failing because it doesn't have xterm installed, as it is the fallback when the system isn't on the list.

That being said, I would like to try to find the user choosen default terminal by other means. On Mate it seems the default terminal is the mate-terminal, even so, I changed it to terminator. 

My proposal is trying to get this information and opening the terminal the user has choosen, instead of the one the system already comes with. And if you could push your refactored code, it could decrease the conflicts I would have if I did it the same time as you.

PS: Sincerely, I don't know how to do that yet, but I am willing to spend some time trying.

**Comment by xovertheyearsx** [comment]
you can do that already, it just hasn't been plugged in to the settings panel. if you don't mind tweaking the code a bit, it should work the way you're expecting it to run. if not, you can always revert to an older [version](https://github.com/foreshadow/atom-python-run/tags) by using one of the archived versions. then use `apm` to do the rest for you.

**A Quick Hack**
There are a few ways to accomplish what you want. The method outlined below is the easiest method and was originally implemented in the base design for the `terminal.js` module. It's a quick hack and solves the issue of the missing settings used to manifest the desired behavior.

Modifying the default system shell was addressed in Issue #32, and unfortunately the wiki hasn't been updated and finished to reflect the latest changes.

In the `atom-python-run.js` source file on line [82](https://github.com/foreshadow/atom-python-run/blob/master/lib/atom-python-run.js#L82) you can insert the code there to modify the terminal object to reflect the desired shell.

**Original Code in `atom-python-run.js`**

```
if ('win32' === process.platform) {
    shell = new terminal.Shell(__dirname + '\\..\\cp\\main.py');
} else {
    shell = new terminal.Shell(__dirname + '/../cp/main.py');
}

spawn = new terminal.SpawnWrapper(shell.object);
// line 82 is here
tty = spawn.tty({
    'pause': config.pause,
    'pipeFile': config.pipeFile,
    'pipePath': config.pipePath,
    'log': config.log,
    'args': args,
    'options': {
        cwd: info.dir,
        detached: true
    }
});
```

**Copy and paste this in to line _#82_**

_Note: this can be anything you want as long as you use the proper flag. here, `-x` is the execution flag for terminator._

```
spawn.object.shell = 'terminator';
spawn.object.option = ['-x'];
```

**Final Result**

```
if ('win32' === process.platform) {
    shell = new terminal.Shell(__dirname + '\\..\\cp\\main.py');
} else {
    shell = new terminal.Shell(__dirname + '/../cp/main.py');
}

spawn = new terminal.SpawnWrapper(shell.object);

spawn.object.shell = 'terminator';
spawn.object.option = ['-x'];

tty = spawn.tty({
    'pause': config.pause,
    'pipeFile': config.pipeFile,
    'pipePath': config.pipePath,
    'log': config.log,
    'args': args,
    'options': {
        cwd: info.dir,
        detached: true
    }
});
```

This allows you to compensate for that gap I couldn't predict. Originally, my notes stated that this was a crude method; I used it to grab only the most common shells. 

When I have some time from studying I can write up a quick patch to add it to the settings panel. I meant to add it in and unfortunately ran out of time. So I pushed what I had to the repo. I honestly wasn't planning on merging it so soon, but that's okay. I can work with it.

If you could write up a patch your self and wouldn't mind posting it here to help out, that would be great as well.
